File: tusimple/哈希/249.group_shifted_strings.py
import collections

class Solution:
    def groupStrings(self, strings):
        
        table = collections.defaultdict(list)
        
        for string in strings:
            translate = '0'
            for i in range(1, len(string)):
                shift = ord(string[i]) - ord(string[i-1])
                if shift < 0:
                    shift = 26 + shift
                translate += str(shift)
            table[(len(string), translate)].append(string)

        return list(table.values())

File: C++/Hash Table/249. Group Shifted Strings.cpp
class Solution {
public:
    vector<vector<string>> groupStrings(vector<string>& strings) {
        unordered_map<string, vector<string>> umap;
        for (string& s : strings) {
            int shift = s[0] - 'a';
            string key = s;
            for (int i = 0; i < s.size(); i++) {
                key[i] = (key[i] - shift >= 'a') ? key[i] - shift : key[i] + 26 - shift;
            }
            umap[key].push_back(s);
        }
        
        vector<vector<string>> res;
        for (auto& item : umap) {
            res.push_back(item.second);
        }
        return res;
    }
};